### ISSUE TITLE
stack overflow in [refine] (confusion about evar dependencies?) (regression since 8.4)

### ISSUE THREAD

**Issue by coqbot** [coq/coq#4631]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4631
From: @JasonGross
Reported version: 8.5
CC:   @gares, @SkySkimmer

**Comment by Armael** [comment]
I just stumbled on what looks like an instance of this issue:

```coq
Goal exists x, x = 0.
  eexists.
  match goal with |- ?x = _ =>
    refine ((fun (_ : x = x) => _) _)
  end.
(* Error: Stack overflow *)
```

Could someone confirm whether this is actually the same issue?

**Comment by Armael** [comment]
Note: the snippet above is derived from an attempt at doing what `eassert` does using `refine`, before finding out that `eassert` exists. However, `eassert` seems to behave correctly, as opposed to refine...

**Comment by maximedenes** [comment]
It seems all these examples are fixed by #7825.
